In [6]:
import pandas as pd

# Exercises

## 1. Given the following confusion matrix, evaluate (by hand) the model's performance.

 |               | actual cat | actual dog |
 |:------------  |-----------:|-----------:|
 | predicted cat |       34 |        7 |  
 | predicted dog |       13 |       46 |  
 
 - In the context of this problem, what is a false positive?

In [23]:
# setting positive to dog
positive = "dog"

tp = 46
tn = 34
fp = 13
fn = 7

accuracy = (tp + tn) / (tp + tn + fp + fn)
recall = tp / (tp + fn)
precision = tp / (tp + fp)

print(f'The model accuracy is: {accuracy: .2%}')
print(f'The model recall is: {recall: .2%}')
print(f'The model precision is: {precision: .2%}')

The model accuracy is:  80.00%
The model recall is:  86.79%
The model precision is:  77.97%


In [28]:
# setting positive to 
positive = "cat"

tp = 34
tn = 46
fp = 7
fn = 13

accuracy = (tp + tn) / (tp + tn + fp + fn)
recall = tp / (tp + fn)
precision = tp / (tp + fp)

print(f'The model accuracy is: {accuracy: .2%}')
print(f'The model recall is: {recall: .2%}')
print(f'The model precision is: {precision: .2%}')

The model accuracy is:  80.00%
The model recall is:  72.34%
The model precision is:  82.93%


- In the context of this problem, what is a false negative?
    - false ative **(FN)** is where we predicted a cat and it was a dog
    

- How would you describe this model?

    - There are two possible predicted classes: `cat` and `dog`
    - The classifier made a total of **_100_** predictions 
    - Out of the 100 cases, the classifier predicted `cat` 34 times, and `dog` 46
    - This model was correct 80% of the time


Accuracy (TP + TN) / (FP + FN + TP + TN)


80 / 100

2. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found [here.](https://ds.codeup.com/data/c3.csv)

Use the predictions dataset and pandas to help answer the following questions:

In [9]:
df = pd.read_csv("c3.csv")
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

positive = defect

negative = no defect

- TP = when we predict a defect and, there is a defect
- FP = when we predict a defect and, there is no defect
- TN = when we predict no defect and , there is no defect
- FN = when we predict a no defect and , there is a defect

In [10]:
pd.crosstab(df.model1, df.actual) # want to reduce FN

actual     Defect  No Defect
model1                      
Defect          8          2
No Defect       8        182

In [11]:
pd.crosstab(df.model2, df.actual)

actual     Defect  No Defect
model2                      
Defect          9         81
No Defect       7        103

In [12]:
pd.crosstab(df.model3, df.actual)

actual     Defect  No Defect
model3                      
Defect         13         86
No Defect       3         98

An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. 

| model 1  |  model 2 | model 3  |
|---|---|---|
| TP = 8  | TP = 9  |  TP = 13 |
| TN = 182  | TN = 103  | TN = 98  |
| FP = 2  | FP = 81  | FP = 86  |
| FN = 8  | FN = 7  | FN = 3  |
|   |   |   |

- Which evaluation metric would be appropriate here? 
## Metrics

- **accuracy**: (TP + TN) / (TP + TN + FP + FN)
    
- **precision**: TP / (TP + FP)
    
    - FP is more costly than FN
- **recall**: TP / (TP + FN)
    
    - FN is more costly than FP

- Which model would be the best fit for this use case?
    - RECALL because missing defects(FN) are more costly in this scenario, than accidentally say something is a defect and it actually isn't

- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

- Model1 = TP / (TP + FN)
8 / (8 + 8) = 0.50 = 50%
- Model2 = TP / (TP + FN)
9 / (9 + 7) = 0.56 = 56%
- Model3 = TP / (TP + FN)
13 / (13 + 3) = 0.81 = 81%

3. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results [here.]( https://ds.codeup.com/data/gives_you_paws.csv)

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

    a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
    
    b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?
    
    c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

- [sklearn.metrics.accuracy_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html)
- [sklearn.metrics.precision_score](sklearn.metrics.precision_score)
- [sklearn.metrics.recall_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.recall_score.html)
- [sklearn.metrics.classification_report](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html)